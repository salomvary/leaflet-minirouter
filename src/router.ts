import * as L from "leaflet";
import polyline from "@mapbox/polyline";

export interface RouteSummary {
  totalTime: number;
  totalDistance: number;
  totalAscend: number;
  totalDescend: number;
}

export interface Route {
  summary?: RouteSummary;
  coordinates: L.LatLng[];
}

export interface Router {
  route(waypoints: L.LatLngExpression[]): Promise<Route[]>;
}

export interface OSRMV1Options {
  serviceUrl?: string;
}

interface OSRMV1Response {
  routes: {
    distance: any;
    duration: any;
    geometry: string;
    legs: {
      steps: {
        geometry: string;
      }[];
    }[];
  }[];
}

const polylinePrecision = 5;

export class OSRMV1 implements Router {
  private serviceUrl: string;
  private profile: string;
  private requestParameters: { [key: string]: string };

  constructor({
    serviceUrl = "https://router.project-osrm.org/route/v1/",
    profile = "driving",
    requestParameters = {},
  } = {}) {
    this.serviceUrl = serviceUrl;
    this.profile = profile;
    this.requestParameters = requestParameters;
  }

  async route(waypoints: L.LatLng[]): Promise<Route[]> {
    const locations = waypoints
      .map(({ lng, lat }) => lng + "," + lat)
      .join(";");
    const url = new URL(this.profile + "/" + locations, this.serviceUrl);
    url.search = new URLSearchParams(this.requestParameters).toString();

    const response: OSRMV1Response = await fetch(url.toString()).then((_) =>
      _.json()
    );

    return response.routes.map((route) => {
      const coordinates = decodePolyline(route.geometry);
      const summary = <RouteSummary>{}; // FIXME
      return { coordinates, summary };
    });
  }
}

export interface GraphHopperOptions {
  serviceUrl?: string;
  requestParameters?: { [key: string]: string };
}

interface GraphHopperResponse {
  paths: Array<{
    points: string;
  }>;
}

/**
 * Router using GraphHopper API:
 * https://docs.graphhopper.com/#operation/getRoute
 */
export class GraphHopper implements Router {
  private apiKey: string;
  private serviceUrl: string;
  requestParameters: { [key: string]: string };

  constructor(
    apiKey: string,
    {
      serviceUrl = "https://graphhopper.com/api/1/route",
      requestParameters = {},
    }: GraphHopperOptions = {}
  ) {
    this.apiKey = apiKey;
    this.serviceUrl = serviceUrl;
    this.requestParameters = requestParameters;
  }

  async route(waypoints: L.LatLng[]): Promise<Route[]> {
    const params = new URLSearchParams({
      ...this.requestParameters,
      key: this.apiKey,
    });
    waypoints.forEach(({ lng, lat }) =>
      params.append("point", lat + "," + lng)
    );
    const url = new URL(this.serviceUrl);
    url.search = params.toString();

    const response: GraphHopperResponse = await fetch(
      url.toString()
    ).then((_) => _.json());

    return response.paths.map((path) => {
      const coordinates = decodePolyline(path.points);
      const summary = <RouteSummary>{}; // FIXME
      return { coordinates, summary };
    });
  }
}

export interface MapboxOptions extends OSRMV1Options {
  profile?: string;
}

export class Mapbox extends OSRMV1 {
  constructor(
    accessToken: string,
    {
      serviceUrl = "https://api.mapbox.com/directions/v5/",
      profile = "mapbox/driving",
    }: MapboxOptions = {}
  ) {
    super({
      serviceUrl,
      profile,
      requestParameters: { access_token: accessToken },
    });
  }
}

function decodePolyline(geometry: string): L.LatLng[] {
  return (polyline.decode(geometry, polylinePrecision) as [
    number,
    number
  ][]).map(L.latLng);
}
