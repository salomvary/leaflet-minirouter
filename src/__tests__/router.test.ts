import * as L from "leaflet";
import { Mapbox, GraphHopper } from "../router";
import mapboxResponse from "./mapbox-response.json";
import graphHopperResponse from "./graphhopper-response.json";

test("Mapbox", async () => {
  window.fetch = jest.fn().mockResolvedValueOnce({
    json: () => Promise.resolve(mapboxResponse),
  });

  const router = new Mapbox("test-token");
  const [route] = await router.route([
    L.latLng(0.1, 0.2),
    L.latLng(-0.3, -0.4),
  ]);

  expect(window.fetch).toHaveBeenCalledWith(
    "https://api.mapbox.com/directions/v5/mapbox/driving/0.2,0.1;-0.4,-0.3?access_token=test-token"
  );
  expect(route.coordinates).toMatchSnapshot();
});

test("GraphHopper", async () => {
  window.fetch = jest.fn().mockResolvedValueOnce({
    json: () => Promise.resolve(graphHopperResponse),
  });

  const router = new GraphHopper("test-token");
  const [route] = await router.route([
    L.latLng(0.1, 0.2),
    L.latLng(-0.3, -0.4),
  ]);

  expect(window.fetch).toHaveBeenCalledWith(
    "https://graphhopper.com/api/1/route?key=test-token&point=0.1%2C0.2&point=-0.3%2C-0.4"
  );
  expect(route.coordinates).toMatchSnapshot();
});
