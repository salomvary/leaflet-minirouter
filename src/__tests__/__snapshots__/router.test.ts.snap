// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphHopper 1`] = `
Array [
  Object {
    "lat": 52.47384,
    "lng": 13.36285,
  },
  Object {
    "lat": 52.47361,
    "lng": 13.36361,
  },
  Object {
    "lat": 52.47262,
    "lng": 13.36584,
  },
  Object {
    "lat": 52.47146,
    "lng": 13.36834,
  },
  Object {
    "lat": 52.47115,
    "lng": 13.3693,
  },
  Object {
    "lat": 52.47093,
    "lng": 13.37029,
  },
  Object {
    "lat": 52.4705,
    "lng": 13.37254,
  },
  Object {
    "lat": 52.47029,
    "lng": 13.37392,
  },
  Object {
    "lat": 52.47017,
    "lng": 13.37483,
  },
  Object {
    "lat": 52.47009,
    "lng": 13.37565,
  },
  Object {
    "lat": 52.46991,
    "lng": 13.37851,
  },
  Object {
    "lat": 52.46987,
    "lng": 13.37992,
  },
  Object {
    "lat": 52.46952,
    "lng": 13.38477,
  },
  Object {
    "lat": 52.46948,
    "lng": 13.38549,
  },
  Object {
    "lat": 52.46948,
    "lng": 13.38574,
  },
  Object {
    "lat": 52.47358,
    "lng": 13.38583,
  },
  Object {
    "lat": 52.47374,
    "lng": 13.38583,
  },
]
`;

exports[`Mapbox 1`] = `
Array [
  Object {
    "lat": 51.50945,
    "lng": -0.13262,
  },
  Object {
    "lat": 51.50807,
    "lng": -0.13138,
  },
  Object {
    "lat": 51.50775,
    "lng": -0.13094,
  },
  Object {
    "lat": 51.50828,
    "lng": -0.12916,
  },
  Object {
    "lat": 51.50761,
    "lng": -0.12874,
  },
  Object {
    "lat": 51.50754,
    "lng": -0.12783,
  },
  Object {
    "lat": 51.50745,
    "lng": -0.12729,
  },
  Object {
    "lat": 51.50652,
    "lng": -0.12239,
  },
  Object {
    "lat": 51.50841,
    "lng": -0.12081,
  },
  Object {
    "lat": 51.50944,
    "lng": -0.11929,
  },
  Object {
    "lat": 51.51026,
    "lng": -0.11693,
  },
  Object {
    "lat": 51.51084,
    "lng": -0.11409,
  },
  Object {
    "lat": 51.51104,
    "lng": -0.11206,
  },
  Object {
    "lat": 51.51099,
    "lng": -0.10281,
  },
  Object {
    "lat": 51.51124,
    "lng": -0.0989,
  },
  Object {
    "lat": 51.51115,
    "lng": -0.09534,
  },
  Object {
    "lat": 51.51085,
    "lng": -0.09302,
  },
  Object {
    "lat": 51.50735,
    "lng": -0.09496,
  },
  Object {
    "lat": 51.50491,
    "lng": -0.09564,
  },
  Object {
    "lat": 51.50476,
    "lng": -0.09396,
  },
  Object {
    "lat": 51.50391,
    "lng": -0.09456,
  },
  Object {
    "lat": 51.50324,
    "lng": -0.09171,
  },
  Object {
    "lat": 51.50377,
    "lng": -0.09127,
  },
  Object {
    "lat": 51.50367,
    "lng": -0.09093,
  },
]
`;
